specter-config/src/client/java/dev/spiritstudios/specter/impl/config/gui/ConfigScreen.java

```diff
@@ -1,39 +1,62 @@
 package dev.spiritstudios.specter.impl.config.gui;
 
+import dev.spiritstudios.specter.api.ConfigScreenManager;
 import dev.spiritstudios.specter.api.config.Config;
-import dev.spiritstudios.specter.api.config.NestedConfig;
-import dev.spiritstudios.specter.api.config.annotations.Range;
 import dev.spiritstudios.specter.api.core.SpecterGlobals;
-import dev.spiritstudios.specter.api.core.util.ReflectionHelper;
-import dev.spiritstudios.specter.impl.config.gui.widget.*;
+import dev.spiritstudios.specter.impl.config.gui.widget.OptionsScrollableWidget;
 import net.minecraft.client.gui.DrawContext;
 import net.minecraft.client.gui.screen.Screen;
 import net.minecraft.client.gui.widget.ButtonWidget;
 import net.minecraft.client.gui.widget.ClickableWidget;
 import net.minecraft.text.Text;
+import net.minecraft.util.Identifier;
 
-import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
+import java.util.function.BiFunction;
 
 public class ConfigScreen extends Screen {
-	private final Config config;
+	private final Config<?> config;
 	private final Screen parent;
 
-	public ConfigScreen(Config config, Screen parent) {
-		super(Text.translatable("config." + config.getId() + ".title"));
+	public ConfigScreen(Config<?> config, Screen parent) {
+		super(Text.translatable("config.%s.%s.title".formatted(config.getId().getNamespace(), config.getId().getPath())));
 		this.config = config;
 		this.parent = parent;
 	}
 
 	@Override
 	protected void init() {
 		super.init();
+		if (this.client == null) close();
 
 		OptionsScrollableWidget scrollableWidget = new OptionsScrollableWidget(this.client, this.width, this.height - 64, 32, 25);
 		List<ClickableWidget> options = new ArrayList<>();
-		for (Field field : config.getClass().getDeclaredFields()) addOptionWidget(field, options);
+
+		List<Config.Value<?>> values = config.getValues().toList();
+
+		boolean failed = false;
+		for (Config.Value<?> option : values) {
+			if (!option.sync() || this.client.player == null || this.client.isInSingleplayer()) continue;
+
+			this.client.player.sendMessage(Text.of("Configs cannot be edited in multiplayer"), true);
+			this.client.setScreen(this.parent);
+			failed = true;
+			break;
+		}
+		if (failed) return;
+
+		values.forEach(option -> {
+			BiFunction<Config.Value<?>, Identifier, ? extends ClickableWidget> factory = ConfigScreenManager.getWidgetFactory(option, this.config.getId());
+			if (factory == null) {
+				SpecterGlobals.LOGGER.warn("No widget factory found for {}", option.defaultValue().getClass().getSimpleName());
+				return;
+			}
+
+			ClickableWidget widget = factory.apply(option, this.config.getId());
+			if (widget != null) options.add(widget);
```

Comment:
Yeah this should probably throw an exception (or atleast log an error). If widget is null, then whoever made the factory messed up. The widgetfactories arent used for builtins (i think i left a comment explaining why), so this would only ever get tripped by a bad implementation by another dev